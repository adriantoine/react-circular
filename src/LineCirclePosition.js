import PropTypes from 'prop-types';
import {toRadians} from './utils';

const LineCirclePosition = props => {
  const radians = toRadians(props.angle);
  const sin = Math.sin(radians);
  const cos = Math.cos(radians);

  const x1 = sin * (props.radius + props.adjust) + props.radius;
  const y1 = cos * (props.radius + props.adjust) + props.radius;
  const x2 = sin * (props.radius + props.adjust + props.length) + props.radius;
  const y2 = cos * (props.radius + props.adjust + props.length) + props.radius;

  const render = props.render || props.children;
  return render({x1, y1, x2, y2});
};

LineCirclePosition.defaultProps = {
  adjust: 0,
  children: () => {},
};

LineCirclePosition.propTypes = {
  angle: PropTypes.number.isRequired,
  radius: PropTypes.number.isRequired,
  adjust: PropTypes.number.isRequired,
  length: PropTypes.number.isRequired,
  render: PropTypes.func,
  children: PropTypes.func,
};

export default LineCirclePosition;
