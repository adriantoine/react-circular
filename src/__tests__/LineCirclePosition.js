/* eslint-env jest */

import React from 'react';
import LineCirclePosition from '../LineCirclePosition';
import Enzyme, {mount} from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';

Enzyme.configure({adapter: new Adapter()});

it('calculates a basic position', () => {
  expect(
    mount(
      <LineCirclePosition angle={90} radius={100} length={20}>
        {({x1, y1, x2, y2}) => <line x1={x1} y1={y1} x2={x2} y2={y2} />}
      </LineCirclePosition>,
    ),
  ).toMatchSnapshot();
});

it('calculates a complex position', () => {
  expect(
    mount(
      <LineCirclePosition angle={41} radius={100} length={20}>
        {({x1, y1, x2, y2}) => <line x1={x1} y1={y1} x2={x2} y2={y2} />}
      </LineCirclePosition>,
    ),
  ).toMatchSnapshot();
});

it('calculates a complex position with an adjust prop', () => {
  expect(
    mount(
      <LineCirclePosition angle={41} radius={100} adjust={-20} length={20}>
        {({x1, y1, x2, y2}) => <line x1={x1} y1={y1} x2={x2} y2={y2} />}
      </LineCirclePosition>,
    ),
  ).toMatchSnapshot();
});

it('uses the render prop when provided', () => {
  expect(
    mount(
      <LineCirclePosition
        angle={90}
        radius={100}
        length={20}
        render={({x1, y1, x2, y2}) => <line x1={x1} y1={y1} x2={x2} y2={y2} />}
      >
        {({x1, y1, x2, y2}) => (
          <text x={x1} y={y1}>
            NOT RENDERED
          </text>
        )}
      </LineCirclePosition>,
    ),
  ).toMatchSnapshot();
});
