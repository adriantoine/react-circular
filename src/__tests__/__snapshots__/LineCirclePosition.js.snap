// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`calculates a basic position 1`] = `
<LineCirclePosition
  adjust={0}
  angle={90}
  length={20}
  radius={100}
>
  <line
    x1={200}
    x2={220}
    y1={100}
    y2={100.00000000000001}
  />
</LineCirclePosition>
`;

exports[`calculates a complex position 1`] = `
<LineCirclePosition
  adjust={0}
  angle={41}
  length={20}
  radius={100}
>
  <line
    x1={165.60590289905073}
    x2={178.7270834788609}
    y1={175.47095802227722}
    y2={190.56514962673265}
  />
</LineCirclePosition>
`;

exports[`calculates a complex position with an adjust prop 1`] = `
<LineCirclePosition
  adjust={-20}
  angle={41}
  length={20}
  radius={100}
>
  <line
    x1={152.48472231924057}
    x2={165.60590289905073}
    y1={160.37676641782176}
    y2={175.47095802227722}
  />
</LineCirclePosition>
`;

exports[`uses the render prop when provided 1`] = `
<LineCirclePosition
  adjust={0}
  angle={90}
  length={20}
  radius={100}
  render={[Function]}
>
  <line
    x1={200}
    x2={220}
    y1={100}
    y2={100.00000000000001}
  />
</LineCirclePosition>
`;
